packages/smart-contracts/src/contracts/BatchConversionPayments.sol

```diff
@@ -0,0 +1,409 @@
+          break;
+        }
+      }
+
+      require(address(this).balance >= amountAndFeeToPay, 'not enough funds');
+
+      // Batch contract pays the requests through EthConversionProxy
+      conversionPaymentEthProxy.transferWithReferenceAndFee{value: amountAndFeeToPay}(
+        payable(rI._recipient),
+        rI._requestAmount,
+        rI._path,
+        rI._paymentReference,
+        rI._feeAmount,
+        _feeAddress,
+        rI._maxRateTimespan
+      );
+    }
+
+    // Check that batch contract has enough funds to pay batch conversion fees
+    uint256 amountBatchFees = ((((contractBalance - address(this).balance) * 10000) /
+      (10000 + basicFee)) * batchConversionFee) / 10000;
+    require(address(this).balance >= amountBatchFees, 'not enough funds for batch conversion fees');
+
+    // Batch contract pays batch fee
+    _feeAddress.transfer(amountBatchFees);
+
+    // Batch contract transfers the remaining ethers to the payer
+    if (address(this).balance > 0) {
+      (bool sendBackSuccess, ) = payable(msg.sender).call{value: address(this).balance}('');
+      require(sendBackSuccess, 'Could not send remaining funds to the payer');
+    }
+  }
+
+  /*
+   * Helper functions
+   */
+
+  /**
+   * @notice Calculate the amount of the conversion
+   */
+  function amountAndFeeConversion(
+    uint256 requestAmount,
+    uint256 requestFee,
+    uint256 rate,
+    uint256 decimals
+  ) private pure returns (uint256) {
+    return (requestAmount * rate) / decimals + (requestFee * rate) / decimals;
+  }
+
+  /**
+   * @notice Get conversion rate and decimals from chainlink
+   */
+  function getRate(address[] memory _path, uint256 _maxRateTimespan)
+    internal
+    view
+    returns (uint256, uint256)
+  {
+    (uint256 rate, uint256 oldestTimestampRate, uint256 decimals) = chainlinkConversionPath.getRate(
+      _path
+    );
+
+    // Check rate timespan
+    require(
+      _maxRateTimespan == 0 || block.timestamp - oldestTimestampRate <= _maxRateTimespan,
+      'aggregator rate is outdated'
+    );
+    return (rate, decimals);
+  }
+
+  /*
+   * Admin functions to edit the conversion proxies address
+   */
+
+  /** fees applied on a single request */
+  function setBasicFee(uint256 _basicFee) public onlyOwner {
+    basicFee = _basicFee;
+  }
+
+  function setBatchConversionFee(uint256 _batchConversionFee) public onlyOwner {
+    batchConversionFee = _batchConversionFee;
+  }
+
+  function setConversionPaymentProxy(address _paymentErc20ConversionFeeProxy) public onlyOwner {
+    conversionPaymentProxy = IERC20ConversionProxy(_paymentErc20ConversionFeeProxy);
+  }
+
+  function setEthConversionPaymentProxy(address _paymentEthConversionFeeProxy) public onlyOwner {
+    conversionPaymentEthProxy = IEthConversionProxy(_paymentEthConversionFeeProxy);
+  }
+
+  /**
+   * @notice Update the conversion path contract used to fetch conversions
+   * @param _chainlinkConversionPathAddress address of the conversion path contract
```

Comment:
Any reason why all the admin functions are public except setConversionPathAddress that is external ?

Reply:
right thx